src/Raven.Client/Http/RavenCommand.cs

```diff
@@ -116,21 +119,67 @@ public virtual async Task<ResponseDisposeHandling> ProcessResponse(JsonOperation
                     // we intentionally don't dispose the reader here, we'll be using it
                     // in the command, any associated memory will be released on context reset
                     var json = await context.ReadForMemoryAsync(stream, "response/object").ConfigureAwait(false);
-                    if (cache != null) //precaution
+
+                    await SetFirstSuccessfulResult(() =>
                     {
-                        CacheResponse(cache, url, response, json);
-                    }
-                    SetResponse(context, json, fromCache: false);
+                        if (cache != null) //precaution
+                        {
+                            CacheResponse(cache, url, response, json);
+                        }
+                        SetResponse(context, json, fromCache: false);
+                    }).ConfigureAwait(false);
+
                     return ResponseDisposeHandling.Automatic;
                 }
 
                 // We do not cache the stream response.
                 using (var uncompressedStream = await RequestExecutor.ReadAsStreamUncompressedAsync(response).ConfigureAwait(false))
-                    SetResponseRaw(response, uncompressedStream, context);
+                {
+                    await SetFirstSuccessfulResult(() =>
+                    {
+                        SetResponseRaw(response, uncompressedStream, context);
+                    }).ConfigureAwait(false);
+                }
             }
             return ResponseDisposeHandling.Automatic;
         }
 
+        private TaskCompletionSource<bool> _successfulResponseWaiter;
+
+        internal void SetTimeout(TimeSpan timeout)
+        {
+            Timeout = timeout;
+        }
+
+        private async Task SetFirstSuccessfulResult(Action action)
```

Comment:
Isn't this method a bit of an overkill? SemaphoreSlim should be enough? It would make this method more readable.

Reply:
done